### PR TITLE
Jxiao/dummy market maker rust

### PR DESCRIPTION
**Pull request by jarry-xiao**

### CODE REVIEW

File: rust/examples/src/bin/simple_market_maker.rs [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+const QUOTE_REFRESH_FREQUENCY: u64 = 2000;
+
+// Edge in bps on quote. Places bid/ask at fair price -/+ edge
+const QUOTE_EDGE_BPS: u64 = 5;
+
+// Size of quote in quote units
+const QUOTE_SIZE_IN_QUOTE_UNITS: f64 = 250.0;
+
+// Expected life time of order in seconds
+const ORDER_LIFETIME_IN_SECONDS: i64 = 7;
```
Comment (eugene-chen) [inline rust/examples/src/bin/simple_market_maker.rs:32]: why not u64?

File: rust/examples/src/bin/simple_market_maker.rs [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+
+        let bid_size = QUOTE_SIZE_IN_QUOTE_UNITS / bid_price;
+        let ask_size = QUOTE_SIZE_IN_QUOTE_UNITS / ask_price;
+
+        let clock_account_data = sdk.client.get_account_data(&sysvar::clock::id()).await?;
+
+        let clock: Clock = bincode::deserialize(&clock_account_data)
+            .map_err(|_| anyhow::Error::msg("Error deserializing clock"))?;
+
+        let bid_order_packet = OrderPacket::Limit {
```
Comment (eugene-chen) [inline rust/examples/src/bin/simple_market_maker.rs:116]: why not use helper method for creating limit orders?

File: rust/examples/src/bin/simple_market_maker.rs [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+            quote_mint,
+            &ask_order_packet,
+        );
+
+        ixs.push(bid_ix);
+        ixs.push(ask_ix);
+
+        let txid = sdk.client.sign_send_instructions(ixs, vec![]).await?;
+
+        println!("Place Txid: {}", txid);
```
Comment (eugene-chen) [inline rust/examples/src/bin/simple_market_maker.rs:175]: add link to explorer?

**Review by eugene-chen — COMMENTED** [review]
Code looks fine. We need to add a TON of documentation here, in the readme, in the code, and/or in the gitbook

* What's the best way to get this running on devnet
* How to use your own RPC
* How to use devnet-token-faucet

Basically want to make it as easy as possible for someone to get this working for the first time, and confirm for themselves that it's working (include links to explorer for successful order placements)

Does the user need a seat for this to work?

File: rust/examples/src/bin/simple_market_maker.rs (hunk 1 above)
Comment (jarry-xiao) [reply rust/examples/src/bin/simple_market_maker.rs:32]: unix timestamp on the clock is i64

File: rust/examples/src/bin/simple_market_maker.rs (hunk 2 above)
Comment (jarry-xiao) [reply rust/examples/src/bin/simple_market_maker.rs:116]: Can change to using a helper. I wanted an example that explicitly states every parameter here

**Comment by jarry-xiao** [comment]
I think we should wait to merge this for now and pull the trigger when permissionless limit order stuff is also 100% ready to go.

That way we can add logic in here that actually claims new seats etc. For now I can comment some of the code

File: rust/examples/src/bin/simple_market_maker.rs (hunk 2 above)
Comment (eugene-chen) [reply rust/examples/src/bin/simple_market_maker.rs:116]: let's put in both and leave one commented out, and both with heavy docs

**Comment by eugene-chen** [comment]
plan sounds good. And we can set a devnet market w/ permissionless limit orders as the default market for this, along with token claiming as the first thing that happens in the script

File: rust/examples/Cargo.toml [hunk 4]
```diff
@@ -25,4 +26,7 @@ shellexpand = { workspace = true }
 borsh = { workspace = true }
 phoenix-v1 = { workspace = true }
 phoenix-sdk = { path = "../phoenix-sdk" }
-devnet-token-faucet = { git = "https://github.com/Ellipsis-Labs/devnet-token-faucet" }
\ No newline at end of file
+devnet-token-faucet = { git = "https://github.com/Ellipsis-Labs/devnet-token-faucet" }
+reqwest = "0.11.14"
+serde_json = "1.0.94"
+bincode = "1.3.3"
```
Comment (eugene-chen) [inline rust/examples/Cargo.toml:32]: please use workspace version

File: rust/examples/src/bin/simple_market_maker.rs [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+        "devnet" | "dev" | "d" => "https://api.devnet.solana.com",
+        "mainnet" | "main" | "m" | "mainnet-beta" => "https://api.mainnet-beta.solana.com",
+        "localnet" | "localhost" | "l" | "local" => "http://localhost:8899",
+        _ => network_str,
+    }
+}
+
+#[tokio::main]
+async fn main() -> anyhow::Result<()> {
+    // Connect to the solana network and get the market address
```
Comment (eugene-chen) [inline rust/examples/src/bin/simple_market_maker.rs:63]: fix doc

File: rust/examples/src/bin/simple_market_maker.rs [hunk 6]
```diff
@@ -0,0 +1,224 @@
...
+    let args = Args::parse();
+    let config = match CONFIG_FILE.as_ref() {
+        Some(config_file) => Config::load(config_file).unwrap_or_else(|_| {
+            println!("Failed to load config file: {}", config_file);
+            Config::default()
+        }),
+        None => Config::default(),
+    };
+    let trader = get_payer_keypair_from_path(&args.keypair_path.unwrap_or(config.keypair_path))?;
```
Comment (eugene-chen) [inline rust/examples/src/bin/simple_market_maker.rs:72]: kinda wack that these override each other

File: rust/examples/src/bin/simple_market_maker.rs [hunk 7]
```diff
@@ -0,0 +1,224 @@
...
+    let market = args.market;
+    sdk.add_market(&market).await?;
+
+    let MarketMetadata {
+        base_mint,
+        quote_mint,
+        ..
+    } = sdk.get_market_metadata(&args.market);
+
+    let mainnet_rpc = RpcClient::new("https://api.mainnet-beta.solana.com".to_string());
```
Comment (eugene-chen) [inline rust/examples/src/bin/simple_market_maker.rs:85]: this overrides the network URL passed in by the user

File: rust/examples/src/bin/simple_market_maker.rs (hunk 6 above)
Comment (jarry-xiao) [reply rust/examples/src/bin/simple_market_maker.rs:72]: would you prefer no override and instead an error?

File: rust/examples/src/bin/simple_market_maker.rs (hunk 7 above)
Comment (jarry-xiao) [reply rust/examples/src/bin/simple_market_maker.rs:85]: This is because we need to use mainnet to get the mSOL account data

Comment (jarry-xiao) [reply rust/examples/src/bin/simple_market_maker.rs:85]: can probably add some logic where we just clone the Arc if the supplied RPC is already mainnet

**Pull request closed without merging**